File: c/examples/query.c
#include <stdio.h>
#include <string.h>

#include <taiwan_lottery/query.h>

#include "../src/game_args.h"

static void print_usage(const char *program) {
    fprintf(stderr, "Usage:\n");
    fprintf(stderr, "  %s local <game> period <PERIOD> [output_dir]\n", program);
    fprintf(stderr, "  %s local <game> month <YYYY-MM> [output_dir]\n", program);
    fprintf(stderr, "  %s local <game> month-range <YYYY-MM> <YYYY-MM> [output_dir]\n", program);
    fprintf(stderr, "  %s remote <game> period <PERIOD>\n", program);
    fprintf(stderr, "  %s remote <game> month <YYYY-MM>\n", program);
    fprintf(stderr, "  %s remote <game> month-range <YYYY-MM> <YYYY-MM>\n", program);
    fprintf(stderr, "  game: super-lotto638 | lotto649 | daily539 | 3d | 4d | 49m6 | 39m5 | 38m6 | 1224 | 740 | tic-tac-toe | 638 | bingo-bingo\n");
}


static void print_numbers(const int32_t *values, size_t len) {
    size_t i;
    printf("[");
    for (i = 0; i < len; ++i) {
        if (i > 0) {
            printf(", ");
        }
        printf("%d", values[i]);
    }
    printf("]");
}

int main(int argc, char **argv) {
    const char *program = argc > 0 ? argv[0] : "query";
    const char *default_output_dir = "data";

    if (argc < 5) {
        print_usage(program);
        return 2;
    }

    const char *source = argv[1];
    int game = taiwan_lottery_parse_game_arg(argv[2]);
    const char *query_mode = argv[3];
    int is_month_range = strcmp(query_mode, "month-range") == 0;
    int value_count = is_month_range ? 2 : 1;
    int status;
    taiwan_lottery_history_draw_page *page = NULL;

    if (game < 0) {
        fprintf(stderr, "Invalid game: %s\n", argv[2]);
        print_usage(program);
        return 2;
    }

    if (strcmp(query_mode, "period") != 0 && strcmp(query_mode, "month") != 0 && !is_month_range) {
        fprintf(stderr, "Query mode must be period, month, or month-range\n");
        print_usage(program);
        return 2;
    }

    if (strcmp(source, "local") == 0) {
        if (argc < 4 + value_count + 1) {
            fprintf(stderr, "Missing query value(s) for mode: %s\n", query_mode);
            print_usage(program);
            return 2;
        }
        const char *output_dir = argc > 4 + value_count ? argv[4 + value_count] : default_output_dir;
        const char *period = strcmp(query_mode, "period") == 0 ? argv[4] : NULL;
        const char *month = !period ? argv[4] : NULL;
        const char *end_month = is_month_range ? argv[5] : month;

        status = query_history_draw(output_dir, game, period, month, end_month, &page);
    } else if (strcmp(source, "remote") == 0) {
        if (argc < 4 + value_count) {
            fprintf(stderr, "Missing query value(s) for mode: %s\n", query_mode);
            print_usage(program);
            return 2;
        }

        const char *period = strcmp(query_mode, "period") == 0 ? argv[4] : NULL;
        const char *month = !period ? argv[4] : NULL;
        const char *end_month = is_month_range ? argv[5] : month;

        status = query_history_draw_from_taiwan_lottery(game, period, month, end_month, &page);
    } else {
        print_usage(program);
        return 2;
    }

    if (status != TAIWAN_LOTTERY_OK) {
        fprintf(stderr, "Query failed (status=%d)\n", status);
        return 1;
    }

    printf("total_size=%zu\n", page->total_size);
    for (size_t i = 0; i < page->item_count; ++i) {
        const taiwan_lottery_history_draw_item *item = &page->items[i];
        printf("period=%s\n", item->period != NULL ? item->period : "");
        printf("date=%s\n", item->date != NULL ? item->date : "");

        printf("numbers=");
        print_numbers(item->numbers.base.numbers, item->numbers.base.numbers_len);
        printf("\n");

        if (item->numbers.sorted_numbers != NULL && item->numbers.sorted_numbers_len > 0) {
            printf("numbers_sorted=");
            print_numbers(item->numbers.sorted_numbers, item->numbers.sorted_numbers_len);
            printf("\n\n");
        } else {
            printf("numbers_sorted=<not available>\n\n");
        }
    }

    free_history_draw_page(page);
    return 0;
}

File: examples/query.rs
use std::path::PathBuf;

use taiwan_lottery::{
    query_history_draw, query_history_draw_from_taiwan_lottery, HistoryDrawQuery, HistoryGame,
};

fn print_usage(program: &str) {
    eprintln!("Usage:");
    eprintln!("  {program} local <game> period <PERIOD> [output_dir]");
    eprintln!("  {program} local <game> month <YYYY-MM> [output_dir]");
    eprintln!("  {program} local <game> month-range <YYYY-MM> <YYYY-MM> [output_dir]");
    eprintln!("  {program} remote <game> period <PERIOD>");
    eprintln!("  {program} remote <game> month <YYYY-MM>");
    eprintln!("  {program} remote <game> month-range <YYYY-MM> <YYYY-MM>");
    eprintln!("  game: super-lotto638 | lotto649 | daily539 | 3d | 4d | 49m6 | 39m5 | 38m6 | 1224 | 740 | tic-tac-toe | 638");
}

fn default_output_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("data")
}

fn parse_game(value: &str) -> Option<HistoryGame> {
    match value.to_ascii_lowercase().as_str() {
        "super-lotto638" | "superlotto638" | "5134" => Some(HistoryGame::SuperLotto638),
        "lotto649" | "5118" => Some(HistoryGame::Lotto649),
        "daily539" | "5120" => Some(HistoryGame::Daily539),
        "3d" | "2108" => Some(HistoryGame::Lotto3D),
        "4d" | "2109" => Some(HistoryGame::Lotto4D),
        "49m6" | "1121" => Some(HistoryGame::Lotto49M6),
        "39m5" | "1197" => Some(HistoryGame::Lotto39M5),
        "38m6" | "5122" => Some(HistoryGame::Lotto38M6),
        "1224" | "5290" => Some(HistoryGame::Lotto1224),
        "740" | "2300" => Some(HistoryGame::Lotto740),
        "tic-tac-toe" | "tictactoe" | "2400" => Some(HistoryGame::TicTacToe),
        "638" | "2500" => Some(HistoryGame::Lotto638),
        _ => None,
    }
}


fn parse_query(mode: &str, args: &[String], value_index: usize) -> Option<HistoryDrawQuery> {
    let value = args.get(value_index)?.clone();
    let query = match mode {
        "period" => HistoryDrawQuery::by_period(value),
        "month-range" => {
            let end_month = args.get(value_index + 1).cloned().unwrap_or_else(|| value.clone());
            HistoryDrawQuery::by_month_range(value, end_month)
        }
        _ => HistoryDrawQuery::by_month(value),
    };
    Some(query)
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let program = args.first().cloned().unwrap_or_else(|| "query".to_string());

    if args.len() < 5 {
        print_usage(&program);
        std::process::exit(2);
    }

    let source = args[1].as_str();
    let Some(game) = parse_game(&args[2]) else {
        eprintln!("Invalid game: {}", args[2]);
        print_usage(&program);
        std::process::exit(2);
    };

    let query_mode = args[3].as_str();
    if query_mode != "period" && query_mode != "month" && query_mode != "month-range" {
        eprintln!("Query mode must be period, month, or month-range");
        print_usage(&program);
        std::process::exit(2);
    }

    // month-range takes two values: <start> <end>, others take one
    let value_count = if query_mode == "month-range" { 2 } else { 1 };

    let result = match source {
        "local" => {
            if args.len() < 4 + value_count {
                eprintln!("Missing query value(s) for mode: {query_mode}");
                print_usage(&program);
                std::process::exit(2);
            }

            let output_dir = args
                .get(4 + value_count)
                .map(PathBuf::from)
                .unwrap_or_else(default_output_dir);
            let Some(query) = parse_query(query_mode, &args, 4) else {
                eprintln!("Missing query value");
                print_usage(&program);
                std::process::exit(2);
            };
            query_history_draw(output_dir, game, query)
        }
        "remote" => {
            if args.len() < 4 + value_count {
                eprintln!("Missing query value(s) for mode: {query_mode}");
                print_usage(&program);
                std::process::exit(2);
            }

            let Some(query) = parse_query(query_mode, &args, 4) else {
                eprintln!("Missing query value");
                print_usage(&program);
                std::process::exit(2);
            };
            query_history_draw_from_taiwan_lottery(game, query)
        }
        _ => {
            print_usage(&program);
            std::process::exit(2);
        }
    };

    match result {
        Ok(page) => {
            println!("total_size={}", page.total_size);
            for item in page.items {
                println!("period={}", item.period);
                println!("date={}", item.date.as_deref().unwrap_or(""));
                println!("numbers={:?}", item.numbers.base.numbers);
                match item.numbers.sorted {
                    Some(numbers_sorted) => println!("numbers_sorted={:?}", numbers_sorted),
                    None => println!("numbers_sorted=<not available>"),
                }
                println!();
            }
        }
        Err(err) => {
            eprintln!("query failed: {err:?}");
            std::process::exit(1);
        }
    }
}
